juju/juju
# cmd/juju/controller: allow registering public controller

This allows a controller with an officially signed certificate to be
registered with:

	juju register <hostname>

Because it's got an officially signed certificate, we can trust it to
obtain the controller UUID, and external user authentication lets us
know what user has been chosen without needing to select one explicitly.

This caused a significant reshuffle of the register command. In
particular, we now ask for the password before asking for
the controller name because we can't get the controller UUID
without the password.  That moves us in the direction of fixing
https://bugs.launchpad.net/juju/+bug/1614010 because we no longer prompt
for the controller name if we know that the controller cannot be stored.

We also update the controller-name input code so that it loops until a
correct controller name is supplied rather than immediately returning
an error.

The tests needed significant change because all the tests that previously
assumed that the controller name was prompted before any interaction
with the server now need a mock server.

To make this a bit more straightforward,  we add a the Prompter types
in cmd/testing which means that the interaction requirements can be
specified more declaratively.  This has the side effect of testing that
the read/write interleaving is exactly as expected.

### QA

Bootstrap a controller with an officially signed certificate and externally configured users:

    juju bootstrap --debug --config 'autocert-dns-name=<your domain name>' --config 'api-port=443' --config identity-url=https://api.jujucharms.com/identity ec2 aws

Grant everyone permission to access the controller and create models:

    juju grant everyone@external addmodel

Then configure the DNS record for your domain name to point to the IP address of the newly
bootstrapped controller.

Then, on a fresh juju client instance (or with the new controller entry removed from controller.yaml and accounts.yaml) run:

    juju register <your domain name>

That should succeed and log in as your USSO username@external.  You should then be able to create models and list models etc.
